Apply `set_read_only()` to child classes of `EditorProperty` elements

TokageItLab opened the pull request:
Fixed #45769.

The `read_only` flag is in `EditorProperty`, so you can use it in the child classes of `EditorProperty`. According to the [description](https://docs.godotengine.org/ja/stable/classes/class_editorproperty.html#class-editorproperty-property-read-only), the `read_only` flag prevents operations in the inspector. However, if you don't set `read_only` for each GUI element implemented in the child class, they can manipulate the value in the inspector even if `read_only == true`.

Currently, it seems that the `read_only` flag is not commonly used, but for example, in #45699, it is necessary that `read_only` works correctly to prevent handling BonePose in RestMode. Also, the `read_only` flag can only be set in C++ currently, but it may be used in GDScript in the future with some way. So I fixed this.

https://user-images.githubusercontent.com/61938263/129532740-8c7fdd61-aad8-407b-a91e-ce834507b924.mov

TokageItLab committed:
Apply set_read_only() to child classes of EditorProperty elements

KoBeWi commented on scene/gui/check_box.cpp:
```diff
@@ -34,9 +34,7 @@
 
 Size2 CheckBox::get_icon_size() const {
 	Ref<Texture2D> checked = Control::get_theme_icon(SNAME("checked"));
-	Ref<Texture2D> checked_disabled = Control::get_theme_icon(SNAME("checked_disabled"));
 	Ref<Texture2D> unchecked = Control::get_theme_icon(SNAME("unchecked"));
-	Ref<Texture2D> unchecked_disabled = Control::get_theme_icon(SNAME("unchecked_disabled"));
```
Instead of removing these lines, it's better to handle them properly below.

TokageItLab replied:
Sure, I fixed it. Since there are so many of them, I changed it to iterate, but is this okay for null checking?
```cpp
Size2 CheckBox::get_icon_size() const {
	Vector<Ref<Texture2D>> icons;
	icons.push_back(Control::get_theme_icon(SNAME("checked")));
	icons.push_back(Control::get_theme_icon(SNAME("unchecked")));
	icons.push_back(Control::get_theme_icon(SNAME("radio_checked")));
	icons.push_back(Control::get_theme_icon(SNAME("radio_unchecked")));
	icons.push_back(Control::get_theme_icon(SNAME("checked_disabled")));
	icons.push_back(Control::get_theme_icon(SNAME("unchecked_disabled")));
	icons.push_back(Control::get_theme_icon(SNAME("radio_checked_disabled")));
	icons.push_back(Control::get_theme_icon(SNAME("radio_unchecked_disabled")));

	Size2 tex_size = Size2(0, 0);
	for (int i = 0; i < icons.size(); ++i) {
		if (icons[i].is_null()) {
			continue;
		}
		tex_size = Size2(MAX(tex_size.width, icons[i]->get_width()), MAX(tex_size.height, icons[i]->get_height()));
	}
	return tex_size;
}
```

KoBeWi replied:
Yeah it's ok. But I'm concerned with performance of this approach. AFAIK pushing to Vector will re-allocate memory and `get_icon_size()` is a method that might be called relatively often. It would be better if you resized the Vector beforehand and assigned the icons for each index. Or you could use LocalVector, but I'm not sure how it works exactly.

TokageItLab replied:
You have a point. It will look redundant, but I reverted to the original style for performance reasons.

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -328,12 +345,24 @@ EditorPropertyTextEnum::EditorPropertyTextEnum() {
 	update_property();
 }
 
 void EditorPropertyPath::_path_pressed() {
+	if (is_read_only()) {
+		return;
+	}
```
Will this ever be called when the property is disabled? The button is disabled, so it can't emit the signal connected to this method.

TokageItLab replied:
I remember that `path_selected` was checked `is_read_only()` because the event fired even after the button was disabled. Rather, check why the event fires and I should probably fix that one.

TokageItLab replied:
I didn't input the value for setup properly when testing, so it seems that another factor was causing the crash. So I think it's OK to omit it.

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -328,12 +345,24 @@ EditorPropertyTextEnum::EditorPropertyTextEnum() {
+	EditorProperty::set_read_only(p_read_only);
+};
+
 void EditorPropertyPath::_path_selected(const String &p_path) {
+	if (is_read_only()) {
+		return;
+	}
```
This can only happen when the property gets disabled while the dialog is visible, which is very unlikely.

TokageItLab replied:
same above

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -628,6 +678,13 @@ EditorPropertyEnum::EditorPropertyEnum() {
 ///////////////////// FLAGS /////////////////////////
 
+void EditorPropertyFlags::set_read_only(bool p_read_only) {
+	for (int i = 0; i < flags.size(); i++) {
```
You can use `for (Checkbox *check : flags)`.

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -936,6 +993,11 @@ class EditorPropertyLayersGrid : public Control {
 	}
 };
 
+void EditorPropertyLayers::set_read_only(bool p_read_only) {
+	button->set_disabled(p_read_only);
+	EditorProperty::set_read_only(p_read_only);
+};
```
This is not enough to disable this control, there's still the grid of clickable buttons.

KoBeWi commented:
Disabled properties should have revert arrow unavailable probably.

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -1198,6 +1275,11 @@ EditorPropertyFloat::EditorPropertyFloat() {
 
 ///////////////////// EASING /////////////////////////
 
+void EditorPropertyEasing::set_read_only(bool p_read_only) {
+	spin->set_read_only(p_read_only);
+	EditorProperty::set_read_only(p_read_only);
+};
```
This doesn't seem to have any effect, the `spin` is still editable after it appears. Might be some bug.

But there is another thing. The curve box is still editable. You should probably ignore the input if the property is readonly. This will also fix the spin slider problem, because it will never appear.

KoBeWi commented on editor/editor_spin_slider.cpp:
```diff
@@ -243,7 +246,7 @@ void EditorSpinSlider::_draw_spin_slider() {
 	int vofs = (size.height - font->get_height(font_size)) / 2 + font->get_ascent(font_size);
 
-	Color fc = get_theme_color(SNAME("font_color"), SNAME("LineEdit"));
+	Color fc = is_read_only() ? get_theme_color(SNAME("font_uneditable_color"), SNAME("LineEdit")) : get_theme_color(SNAME("font_color"), SNAME("LineEdit"));
```
```suggestion
	Color fc = get_theme_color(is_read_only() ? SNAME("font_uneditable_color") : SNAME("font_color"), SNAME("LineEdit"));
```
Just like above.

TokageItLab commented:
The revert button is now hidden when `is_read_only()==true`.

As for LayerGrid and Ease, the input processing is implemented in `editor_properties.cpp`, soI change the input processing is controlled by the value of `is_read_only()`.

> Also I don't see any way to disable selected properties, only whole inspector can be disabled. This would be useful.

I think it is necessary to separate methods such as `is_all_read_only()` and `is_read_only(idx)` in order to control them separately. What do you think?

The rest is a matter of design. As for Easing, I used `font_uneditable_color` for the number value color, but the line color is not changed. The base color of the LayerGrid should also be changed. I think it is debatable.

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -712,6 +764,14 @@ class EditorPropertyLayersGrid : public Control {
+		read_only = p_read_only;
+	}
+
+	bool is_read_only() {
+		return read_only;
+	}
```
This method is not necessary, as the readonly is checked only internally in this class. Thus you could just do `if (read_only)`.

KoBeWi commented on editor/editor_spin_slider.cpp:
```diff
@@ -222,6 +222,9 @@ void EditorSpinSlider::_draw_spin_slider() {
 	Vector2 size = get_size();
 
 	Ref<StyleBox> sb = get_theme_stylebox(SNAME("normal"), SNAME("LineEdit"));
+	if (is_read_only()) {
+		sb = get_theme_stylebox(SNAME("read_only"), SNAME("LineEdit"));
+	}
```
Why didn't you use `? :` like everywhere else?

KoBeWi commented:
> I think it is necessary to separate methods such as is_all_read_only() and is_read_only(idx) (also set_read_only()) in order to control them separately. What do you think about it?

After second though, `set_read_only()` is the method to disable property, bruh. The problem is that the inspector will overwrite all properties' readonly state inside `update_tree()`. The only solution I can think of is for EditorProperty to have a "global" readonly flag (set by inspector) and internal one. The property would be readonly if either of them is true.

KoBeWi commented on doc/classes/EditorProperty.xml:
```diff
@@ -50,13 +50,34 @@
+		<method name="is_read_only">
+			<return type="bool" />
+			<description>
+				If either [code]global_read_only[/code] or [code]read_only[/code] is [code]true[/code], returns [code]true[/code].
```
```suggestion
				Returns [code]true[/code] if either [code]global_read_only[/code] or [code]read_only[/code] is [code]true[/code].
```
(this is more consistent with how such methods are usually documented)

KoBeWi commented on doc/classes/EditorProperty.xml:
```diff
@@ -50,13 +50,34 @@
+			<argument index="0" name="read_only" type="bool" />
+			<description>
+				Used by the inspector, set to [code]true[/code] when the property is read-only.
+				Note: The inspector will set all property to readonly by iterating, so if you want to set property to readonly individually, use [method set_read_only].
```
```suggestion
				[b]Note[/b]: The inspector will set all property to readonly by iterating, so if you want to set property to readonly individually, use [method set_read_only].
```

TokageItLab commented:
@KoBeWi I got it right, I was mislead to think that you want to readonly `x` in `Vector3` for example yesterday.
After rethinking, I've implemented `global_read_only` for the problem of the inspector overwriting all properties in update, please check if I'm understanding it correctly. So, I've split the commit for that, but I will squash it in final.

KoBeWi commented:
Yeah, it's good now.

Although the `_set_read_only()` method being empty is rather meh. We could bind a virtual method  (with `BIND_VMETHOD`) and call it:
```C++
void EditorProperty::_set_read_only(bool p_read_only) {
	if (get_script_instance() && get_script_instance()->has_method("_set_read_only")) {
		get_script_instance()->call("_set_read_only", p_type);
	}
}
```
This way the behavior could be customized from plugins. (and it also means that you'd need to restore the base method calls...)

Also the method should be `protected`.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -901,6 +913,8 @@ void EditorProperty::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("set_label", "text"), &EditorProperty::set_label);
 	ClassDB::bind_method(D_METHOD("get_label"), &EditorProperty::get_label);
 
+	ClassDB::bind_method(D_METHOD("_set_read_only", "read_only"), &EditorProperty::_set_read_only);
```
It should be
`BIND_VMETHOD(MethodInfo("_set_read_only", PropertyInfo(Variant::BOOL, "readonly"))`
and after all methods.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -901,6 +913,8 @@ void EditorProperty::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_label"), &EditorProperty::get_label);
 
+	ClassDB::bind_method(D_METHOD("_set_read_only", "read_only"), &EditorProperty::_set_read_only);
+	ClassDB::bind_method(D_METHOD("set_global_read_only", "read_only"), &EditorProperty::set_global_read_only);
```
This method is meant to be used only by the inspector, so I'd remove the binding.

KoBeWi commented on editor/editor_properties.h:
```diff
@@ -62,6 +62,7 @@ class EditorPropertyText : public EditorProperty {
 	static void _bind_methods();
 
 public:
+	virtual void _set_read_only(bool p_read_only) override;
```
This should be in `protected` field (next to `_bind_methods()`). Same in all classes.

KoBeWi commented on doc/classes/EditorProperty.xml:
```diff
@@ -50,13 +57,26 @@
+			<return type="void" />
+			<argument index="0" name="read_only" type="bool" />
+			<description>
+				Set the property to read-only with calling [method _set_read_only].
```
I'd change it to something like
```suggestion
				If [code]true[/code], the property can't be edited.
```

KoBeWi commented on doc/classes/EditorProperty.xml:
```diff
@@ -9,6 +9,13 @@
+			<return type="void" />
+			<argument index="0" name="readonly" type="bool" />
+			<description>
+				This virtual function make all items uneditable which may be manipulated by the user in the inspector.
```
```suggestion
				Called when the property's [code]read_only[/code] state changes. Override this method to disable or enable underlying controls.
```

KoBeWi reviewed (approved):
Seems all good now. Though needs a review from @groud before merging.

TokageItLab commented:
@KoBeWi Thank you for all your help!

groud commented on editor/editor_inspector.cpp:
```diff
@@ -382,12 +382,24 @@ void EditorProperty::update_property() {
 	GDVIRTUAL_CALL(_update_property);
 }
 
+void EditorProperty::_set_read_only(bool p_read_only) {
```
I think this is pointless. I believe it might make sense to have this as the base implementation (in the EditorProperty class itself), but there no point in having it in all children classes. You should simply override `set_read_only(bool p_read_only)` on the children classes. 

Most functions in the engine don't check for scripts for performance reason, we only expose the ones that have a clear use case.

TokageItLab replied:
Does this mean that these are meaningless?
https://github.com/godotengine/godot/pull/51722#issuecomment-903175425

groud replied:
I guess. I think allowing custom behavior makes sense for the base class (most, if not all, scripts will inherit from EditorProperty directly) but not for the children one.

groud commented on editor/editor_inspector.cpp:
```diff
@@ -382,12 +382,24 @@ void EditorProperty::update_property() {
+	}
+}
+
+void EditorProperty::set_global_read_only(bool p_read_only) {
```
No, this does not make a lot of sense, it's not up the EditorProperty class to handle something that's part of the EditorInspector.

It's up to the EditorInspector itself to handle this, setting all children EditorProperty to read_only when the inspector itself is set to read_only, or only the correct ones when it's not.

TokageItLab replied:
If there is only one flag, it is not possible to decide whether it is readonly by the inspector or property itself. For example, in the use case #30623, some properties will be readonly individually, but when the inspector becomes `readonly==true` due to other factors, and then the inspector itself becomes `readonly==false`, the readonly flag of all properties is set to `false`, so the individual property's readonly becomes meaningless.

groud replied:
> but when the inspector becomes readonly==true due to other factors, and then the inspector itself becomes readonly==false, the readonly flag of all properties is set to false, so the individual property's readonly becomes meaningless.

No, that's not how it should work. When changing the inspector's read_only status, you can simply re-create the whole tree to solve the problem by calling `update_tree()`. It's not the most efficient way but performance here is not a problem (changing the inspector read_only status is not something done very often). `update_tree()` should then set the property to read_only when either the property is set to read only or if the inspector is set to read_only.

TokageItLab commented:
@groud In other words, the only thing we need to implement here is a virtual function called `set_uneditable()`, for example, and in `update_tree()`
```cpp
if (EditorProperty->is_readonly() || EditorInspector->is_readonly()) {
  EditorProperty->set_uneditable();
}
```
Does this mean that I should do this?

groud commented:
> @groud In other words, the only thing we need to implement here is a virtual function called set_uneditable(), for example, and in update_tree()
if (EditorProperty->is_readonly() || EditorInspector->is_readonly()) {
  EditorProperty->set_uneditable();
}
should I do this, right?

Not really, no.

To make it clear, this PR has a lot of collisions with https://github.com/godotengine/godot/pull/30623. That PR adds a PROPERTY_USAGE_READ_ONLY flag that is used to determine if a property is read_only or not. You should add it in this PR too, and then, when building the tree, you should do something like:
```
bool property_read_only = (p.usage & PROPERTY_USAGE_CATEGORY) || read_only;
...
ep->set_read_only(property_read_only);
```

TokageItLab commented:
@groud `PROPERTY_USAGE_READ_ONLY` flag in #30623 is implemented in the `Control` object, and the flag is implemented with hardcoding. With that implementation, I think it cannot to make it readonly individually by referring to a child class of `EditorProperty` from outside. As use case
> in #45699, it is necessary that read_only works correctly to prevent handling BonePose in RestMode.

groud commented:
> @groud PROPERTY_USAGE_READ_ONLY flag in #30623 is implemented in the Control object, and the flag is implemented with hardcoding and private.

I am not sure what you mean, the flag is added to `object.h, so that any object can set its property to read_only if needed.
BTW, the `_validate_property(PropertyInfo &property)` function is here for that, it is called when listing the object properties so that you can dynamically update the flags of a property. (depending on a property's value, some fields might be disabled for example).

> With that implementation, I think it cannot to make it readonly individually by referring to a child class of EditorProperty from outside.

I don't think why you would need that anyway. It's not a reliable way to identify EditorProperty nodes (at least in the inspector). If you need something more complex than using the default way the inspector does things (by creating EditorProperties out of the list of property), then you need to code the logic inside an inspector plugin. Otherwise, simply updating the nodes properties with the correct flags (in `_validate_property(PropertyInfo &property)`) should be fine.

TokageItLab commented:
@groud Thank you for the explanation, now I understood it somehow. For example, the current `SkeletonGizmo` needs to be set to readonly by a specific edit mode instead of specific object state. In that case, it should implement `_validate_property()` in `skeleton_3d.cpp` to get the edit mode and set the flag, is it correct?

Fixed:
- The `_set_read_only()` remains as a virtual function to control each handling in child classes, but it no longer to call the base method.
- `global_read_only` has been omitted.
- `PROPERTY_USAGE_READ_ONLY` flag is now implemented, and the inspector determines from the object property and the inspector whether it is readonly or not, and controls the handling with `EditorProperty`.

groud commented:
> In that case, it should implement _validate_property() in skeleton_3d.cpp to get the edit mode and set the flag, is it correct?

It seems correct. I am not very familiar with how `skeleton_3d.cpp` work, but if you have a way to know in which mode you are editing (I guess it would set a property on Skeleton3D, maybe the "pose" one ?). Then it should be quite easy to implement.

Otherwise you may add a setter for a new property within Skeleton3D, that would only impact whether or not part of its properties are read-only.

groud commented on editor/editor_inspector.cpp:
```diff
@@ -1954,7 +1962,8 @@ void EditorInspector::update_tree() {
 					ep->set_keying(keying);
 
-					ep->set_read_only(read_only);
+					bool property_read_only = (p.usage & PROPERTY_USAGE_READ_ONLY) || read_only;
```
You can move this out of the `for (Ref<EditorInspectorPlugin> &ped : valid_plugins)` loop, similar to how it's done for `checked` and `checkable`

groud reviewed (approved):
Alright, LGTM

groud commented:
Thanks!
